### PR TITLE
NGINX page added.

### PR DESCRIPTION
**Pull request by m-saxemberg**

### CODE REVIEW

**Comment by Swader** [comment]
@wpank can you look this over? Or if you have someone of a more devopsy mind available, could you ping them?

**Comment by laboon** [comment]
@wpank is this still something we want to include?

**Comment by salmad3** [comment]
Any possible status on this?:)

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
+---
+id: maintain-guides-how-to-setup-a-Validator-with-NGINX-reverse-proxy
+title: Set Up a Validator with NGINX Reverse Proxy
+sidebar_label: Set Up a Validator with NGINX Reverse Proxy
+---
+
+This guide assumes you have already the hardware ready with the appropriate specs. It has same configuration of the
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
This guide assumes that you have already configured your hardware with the appropriate specs. It has same configuration as the
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+[polkadot secure validator](https://github.com/w3f/polkadot-secure-validator).
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
[polkadot validator setup](https://github.com/w3f/polkadot-secure-validator).
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+In this guide, we will walk you through how to configure a reverse proxy in front of your validator. The
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
We will walk you through how to configure a reverse proxy using NGINX in front of your validator node. The
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+validator uses the NGINX reverse proxy only to talk to the public facing nodes to protect it from certain types of attacks.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
validator uses the reverse proxy only to talk to the public facing nodes, to protect it from certain types of attacks.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+### 1. Firewall configuration
+
+We will configure the firewall with ufw. There needs to be three main ports for this setup. The SSH port commonly is 22 which must be allowed, the proxy port which also must be allowed and the p2p port which in turn must be denied at the firewall level. In this example we will assign the port number `2435` to the proxy port and the port number `30333` to the p2p port.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
We will configure the firewall with [ufw](https://wiki.ubuntu.com/UncomplicatedFirewall). There needs to be three main ports for this setup. 
- An SSH port, commonly ssh/tcp port `22`.
- A proxy port 
- p2p port: must be denied at the firewall level. 

In this example, we will assign the port number `2435` to the proxy port and the port number `30333` to the p2p port. To enable the firewall and the use of the ports, allow SSH access.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+ufw allow 22/tcp
+# proxy port
+ufw allow 2435/tcp
+# libp2p port
+ufw deny 30333/tcp
+ufw enable
+# double check the firewall rules
+ufw verbose
+```
+
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md:35]:
```suggestion
The `verbose` option shows some extra information about the firewall's behavior. 
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+### 2. Configure journald logs
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
### 2. Basic log viewing
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+Create a file called `journald.conf` file inside the `/etc/systemd/` directory with the following content:
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md:37]:
```suggestion
We use [journald](https://www.loggly.com/blog/why-journald/) logs for basic log viewing.
Create a file called `journald.conf` file inside the `/etc/systemd/` directory with the following content:
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+RateLimitIntervalSec=30s
+RateLimitBurst=20000
+SystemMaxUse=50G
+SystemMaxFileSize=512M
+SystemMaxFiles=100
+```
+
+Check out the [example journald configuration file](https://github.com/w3f/polkadot-secure-validator/blob/master/ansible/roles/polkadot-validator/files/journald.conf) for more available options.
+
+finally, run the following command to restart the journald service.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
Finally, run the following command to restart the journald service:
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+```bash
+systemctl restart systemd-journald
+```
+
+### 3. NGINX reverse proxy setup
+
+First install NGINX with the following command.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
First, install NGINX with the following command:
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+```
+apt install nginx
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
sudo apt-get install nginx
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+```
+
+Next, create an NGINX configuration file called `nginx.conf` file inside the `/etc/nginx/` directory with the following content.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
Next, create an NGINX configuration file called `nginx.conf` inside the `/etc/nginx/` directory with the following content:
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+http {
+  include sites-enabled/*;
+}
+
+events{
+}
+```
+
+This will import and make use of the [NGINX stream module](https://nginx.org/en/docs/stream/ngx_stream_core_module.html). In a nutshell, this module will allow continous streaming of data in or out of the validator machine with all the benefits of having an optimized reverse proxy.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
This will import and make use of the [NGINX stream module](https://nginx.org/en/docs/stream/ngx_stream_core_module.html). In a nutshell, this module allows for continuous streaming of data in or out of the validator machine with all the benefits of having an optimized reverse proxy.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+Next, create a folder called `/streams-enabled/` inside the `/etc/nginx/` directory and remove the default NGINX site.
+
+```bash
+# Create the streams-enabled folder
+mkdir /etc/nginx/streams-enabled
+# Delete the default configuration
+/etc/nginx/sites-enabled/default
+```
+
+Now, inside the newly created directory `/etc/nginx/streams-enabled/`, create the proxy service file called `polkadot-proxy.conf` with the following content and the previously defined ports. The port number `2435` to the proxy port and the port number `30333` to the p2p port.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
Now, inside the newly created directory `/etc/nginx/streams-enabled/`, create the proxy service file called `polkadot-proxy.conf` with the following content: 
> Use the previously defined ports: port `2435` for the proxy port & port number `30333` for the p2p port.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+}
+```
+
+Change the permissions of the file `polkadot-proxy.conf` accordingly.
+
+```bash
+chmod 0600 polkadot-proxy.conf
+```
+
+Finally, restart NGINX with the following command
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
Finally, restart NGINX with the following command:
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+Now, inside the newly created directory `/etc/nginx/streams-enabled/`, create the proxy service file called `polkadot-proxy.conf` with the following content and the previously defined ports. The port number `2435` to the proxy port and the port number `30333` to the p2p port.
+
+```bash
+server {
+  listen 0.0.0.0:2435;
+  proxy_pass localhost:30333;
+}
+```
+
+Change the permissions of the file `polkadot-proxy.conf` accordingly.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
Change the permissions of the file `polkadot-proxy.conf` accordingly:
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+
+Finally, restart NGINX with the following command
+
+```bash
+service nginx restart
+```
+
+### 4. Defining your proxy port and p2p port in the polkadot command
+
+These are some of the flags you are going to use.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
These are some of the flags you are going to use when executing the command.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+`--public-addr <VALIDATOR_IP>, <PROXY_PORT>` - This flag defines the validator's IP and the proxy port
+that all other nodes in the network will connect to.
+
+`--listen-addr <LOCALHOST>, <P2P_PORT>` - This flag defines the p2p port that the polkadot application
+will use to connect to the NGINX reverse proxy.
+
+#### P2P Networking
+
+Nodes will use [libp2p](https://libp2p.io/) as the networking layer to establish peers and gossip
+messages but will need to go through NGINX first as a first listener of the streaming data to help balance the load.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
messages, but uses NGINX as a load balancer which acts as a *first listener* of the streaming data to help balance the load.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+
+##### public-addr
+
+`public-addr` - A `public-addr` is a flexible encoding of multiple layers of protocols into a human
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
`public-addr` - a flexible encoding of multiple layers of protocols into a human
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+readable addressing scheme. In our example, `/ip4/<VALIDATOR_IP>/tcp/<PROXY_PORT>` is a valid `public-addr` that
+specifies you want the network to reach the Validator IPv4 address with tcp packets on the predefined proxy port.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
specifies wanting the network to reach the validator IPv4 address with TCP packets on the pre-defined proxy port.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+
+- `IP_ADDRESS` - The public IP address of the validator.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
- `IP_ADDRESS` - the public IP address of the validator.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+
+- `PROXY_PORT` - This is the port that nodes will send p2p messages over the network and will be read by the NGINX reverse proxy.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
- `PROXY_PORT` - the port that nodes will send p2p messages over the network and are read by the NGINX reverse proxy.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+
+##### listen-addr
+
+`listen-addr` - A `public-addr` is the specification of what port the polkadot application will connect to the reverse proxy.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
`listen-addr` -  the specification of what port the polkadot application will connect to the reverse proxy.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+In our example, `/ip4/127.0.0.1/tcp/<P2P_PORT>` means that
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
In our example, `/ip4/127.0.0.1/tcp/<P2P_PORT>` 
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+specifies you want to listen to NGINX on the localhost address `127.0.0.1` with tcp packets on the predefined p2p port.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
specifies that you want to listen to NGINX on the localhost address (`127.0.0.1`), with TCP packets on the pre-defined p2p port.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+
+- `P2P_PORT` - This is the port that the polkadot application will connect to NGINX.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
- `P2P_PORT` - the port that the polkadot application connects to NGINX.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+
+#### Starting Validator with the NGINX Proxy
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
#### Starting the validator with the NGINX proxy
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 28]
```diff
@@ -0,0 +1,176 @@
...
+
+After retrieving the appropriate `IP_ADDRESS`, `PROXY_PORT` and `P2P_PORT` of the validator node, we can start the validator
+using the following flags:
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion

```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 29]
```diff
@@ -0,0 +1,176 @@
...
+
+`listen-addr` - A `public-addr` is the specification of what port the polkadot application will connect to the reverse proxy.
+In our example, `/ip4/127.0.0.1/tcp/<P2P_PORT>` means that
+specifies you want to listen to NGINX on the localhost address `127.0.0.1` with tcp packets on the predefined p2p port.
+
+- `P2P_PORT` - This is the port that the polkadot application will connect to NGINX.
+
+#### Starting Validator with the NGINX Proxy
+
+After retrieving the appropriate `IP_ADDRESS`, `PROXY_PORT` and `P2P_PORT` of the validator node, we can start the validator
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
After retrieving the appropriate `IP_ADDRESS`, `PROXY_PORT` and `P2P_PORT` of the validator node, we can start the validator.
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 30]
```diff
@@ -0,0 +1,176 @@
...
+  --name My_Validator_Name \
+  --validator \
+  --public-addr=/ip4/IP_ADDRESS/tcp/2435 \
+  --listen-addr=/ip4/127.0.0.1/tcp/30333 \
+  --rpc-methods=Unsafe \
+  --chain=Polkadot
+
+```
+
+You should see your validator's peers, and the p2p port you are using to connect to NGINX.
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
You should see your validator's peers, as well as the p2p port you are using to connect to NGINX.
```

**Comment by salmad3** [comment]
From my understanding the configuration hits the mark.
- Should do a complete end-to-end test: consider corner cases; try to break the setup, etc... 
- Have someone with more devops experience review as was mentioned some time ago:)

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 31]
```diff
@@ -0,0 +1,176 @@
+---
+id: maintain-guides-how-to-setup-a-Validator-with-NGINX-reverse-proxy
```
Comment (salmad3) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md]:
```suggestion
id: maintain-guides-how-to-setup-a-validator-with-reverse-proxy
```

File: docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md [hunk 32]
```diff
@@ -0,0 +1,176 @@
...
+We will configure the firewall with ufw. There needs to be three main ports for this setup. The SSH port commonly is 22 which must be allowed, the proxy port which also must be allowed and the p2p port which in turn must be denied at the firewall level. In this example we will assign the port number `2435` to the proxy port and the port number `30333` to the p2p port.
+
+```bash
+# ssh port
+ufw allow 22/tcp
+# proxy port
+ufw allow 2435/tcp
+# libp2p port
+ufw deny 30333/tcp
+ufw enable
```
Comment (jakehemmerle) [inline docs/how-to-setup-a-validator-with-nginx-reverse-proxy.md:29]: might want to put `ufw enable` before the rule additions and then `ufw reload` after. If ufw is running already, I don't think `enable` will reload the rules.

**Review by jakehemmerle — COMMENTED** [review]
@wpank what is the point of having a reverse proxy here? it doesn't load balance anything, authenticate anything, it just passes all requests through another port. I know nginx can do logging etc, but as it is now, it doesn't seem to do anything functional. am I missing something?

**Comment by salmad3** [comment]
> @wpank what is the point of having a reverse proxy here? it doesn't load balance anything, authenticate anything, it just passes all requests through another port. I know nginx can do logging etc, but as it is now, it doesn't seem to do anything functional. am I missing something?

In this configuration, isn't it just being used as an additional layer, and sits behind ufw?

**Comment by jakehemmerle** [comment]
> > @wpank what is the point of having a reverse proxy here? it doesn't load balance anything, authenticate anything, it just passes all requests through another port. I know nginx can do logging etc, but as it is now, it doesn't seem to do anything functional. am I missing something?
> 
> In this configuration, isn't it just being used as an additional layer, and sits behind ufw?

thats my understanding, hence my confusion

**Comment by salmad3** [comment]
Since the PR was opened from a fork, and it seems we should finish off the contribution, I had rebased this to another branch `nginx`. This will also allow the changes to run the latest CI. 

Review can continue there.

**Comment by vcastellm** [comment]
> > > @wpank what is the point of having a reverse proxy here? it doesn't load balance anything, authenticate anything, it just passes all requests through another port. I know nginx can do logging etc, but as it is now, it doesn't seem to do anything functional. am I missing something?
> > 
> > 
> > In this configuration, isn't it just being used as an additional layer, and sits behind ufw?
> 
> thats my understanding, hence my confusion

I've the same concern as @jakehemmerle expressed, this is not doing anything functional, how is this improving the setup?
